pkg/controllers/monitoring/monitoring-stack/conditions.go

```diff
@@ -9,36 +9,45 @@ import (
 )
 
 const (
-	AvailableReason                = "MonitoringStackAvailable"
-	ReconciledReason               = "MonitoringStackReconciled"
-	FailedToReconcileReason        = "FailedToReconcile"
-	PrometheusNotAvailable         = "PrometheusNotAvailable"
-	PrometheusNotReconciled        = "PrometheusNotReconciled"
-	PrometheusDegraded             = "PrometheusDegraded"
-	ResourceSelectorIsNil          = "ResourceSelectorNil"
-	CannotReadPrometheusConditions = "Cannot read Prometheus status conditions"
-	AvailableMessage               = "Monitoring Stack is available"
-	SuccessfullyReconciledMessage  = "Monitoring Stack is successfully reconciled"
-	ResourceSelectorIsNilMessage   = "No resources will be discovered, ResourceSelector is nil"
-	ResourceDiscoveryOnMessage     = "Resource discovery is operational"
-	NoReason                       = "None"
+	AlertmanagerNotAvailable         = "AlertmanagerNotAvailable"
+	AlertmanagerDegraded             = "AlertmanagerDegraded"
+	AlertmanagerNotReconciled        = "AlertmanagerNotReconciled"
+	AvailableReason                  = "MonitoringStackAvailable"
+	ReconciledReason                 = "MonitoringStackReconciled"
+	FailedToReconcileReason          = "FailedToReconcile"
+	PrometheusNotAvailable           = "PrometheusNotAvailable"
+	PrometheusNotReconciled          = "PrometheusNotReconciled"
+	PrometheusDegraded               = "PrometheusDegraded"
+	ResourceSelectorIsNil            = "ResourceSelectorNil"
+	CannotReadPrometheusConditions   = "Cannot read Prometheus status conditions"
+	CannotReadAlertmanagerConditions = "Cannot read Alertmanager status conditions"
+	AvailableMessage                 = "Monitoring Stack is available"
+	SuccessfullyReconciledMessage    = "Monitoring Stack is successfully reconciled"
+	ResourceSelectorIsNilMessage     = "No resources will be discovered, ResourceSelector is nil"
+	ResourceDiscoveryOnMessage       = "Resource discovery is operational"
+	NoReason                         = "None"
 )
 
-func updateConditions(ms *v1alpha1.MonitoringStack, prom monv1.Prometheus, recError error) []v1alpha1.Condition {
+func updateConditions(ms *v1alpha1.MonitoringStack, prom monv1.Prometheus, alertManager *monv1.Alertmanager, recError error) []v1alpha1.Condition {
 	return []v1alpha1.Condition{
 		updateResourceDiscovery(ms),
-		updateAvailable(ms.Status.Conditions, prom, ms.Generation),
-		updateReconciled(ms.Status.Conditions, prom, ms.Generation, recError),
+		updateAvailable(ms.Status.Conditions, prom, alertManager, ms.Generation),
+		updateReconciled(ms.Status.Conditions, prom, alertManager, ms.Generation, recError),
```

Comment:
I feel if possible we should improve this pattern so something on the lines of 

```
conditions := ms.Status.Conditions
conditions = updateResourceDiscovery(conditions, ms)
conditions = updateAlertManagerCondition(conditions, alertmanager, ms.Generation)
conditions = updatePrometheusCondition(conditions, promethues, ms.Generation) 
conditions = updateReconcilationError(conditions, recError)

return conditions
```

I think we can generalise this further by using a condition-setter interface but the  idea is that every step takes as input the previous conditions and decides to update if the severity is higher than existing and return back the updated condition.

WDYT?

Reply:
What do we mean by "update if the severity is higher than existing" that's the only part that I'm concerned about 👀 How would we codify that?

Reply:
Say Monitoring stack has 3 components (or 5 some time in future) 

- prometheus 
- alertmanager (optional)
-  thanos (optional)

We need to update the condition of the stack by reading each of the component's status and update 
* Available
* Reconciled

Currently we pass updateAvailable(prometheus, am, thanos) where the func is somewhat like 
```
func updateAvailable(pm, am, t) {
  if am {
      updateAMAvailable(...)
   }
  if t {
     updateThanosAvailable(...)
  }
  updatePrometheusAvailable
}
```
Say AM is a Degraded, thanos is Unavailable, and Prometheus is Degraded.  What should the status of Stack be? 
I see 2 issues with this. The function takes the responsibility of computing the `available` condition based on all 3 components and it is unclear what comes out of it if you look at the code above. Adding a new component means modifying this (already) huge function. 

Instead of this, I think if you followed the pattern above of passing previous conditions to the next in chain, then you would get something like 
```
conditions := emptyConditions()
conditions = updateResourceDiscovery(conditions, ms) 
conditions = updateThanosCondition(conditions, thanos, ms.Generation) 
conditions = updateAlertManagerCondition(conditions, alertmanager, ms.Generation) 
conditions = updatePrometheusCondition(conditions, promethues, ms.Generation) 
conditions = updateReconcilationError(conditions, recError)
```
Each component updates its condition based on previous components conditions (passed in as input) and the precedence here is that `Prometheus > AlertManager > Thanos`. Each of these methods would look at the input `condition` and overwrite it based on `NotReconciled < Degraded < Unavailable`. 

I will modify this PR a bit to demonstrate what I mean, further.

Reply:
Thanks for thorough description @sthaha. I agree in general but I think it would be good to define all the new states and coresponding conditions first or if everyone is OK with the precedence `Prometheus > AlertManager > Thanos` then fine, but what we will do when there's a new component? Will it be always at the end of the "chain"?

Reply:
@tremes , indeed! I think as such this pr adds value. The refactor (which I am trying out) can follow later. 
Lets go ahead with shipping the feature since I don't expect anything to change due to refactor.

Reply:
> Will it be always at the end of the "chain"?

Yes, the last one has the say.  If an `updateX(..)` needs to know the condition set by `updateY` before, then those need to be explictly passed in as `updateX(..., conditionsSetByUpdateY)` and that makes the dependencies explicit.